Header/Menu improvement

fairlighteth opened the pull request:
# Summary

- Final merge PR to merge in the waterfall PR series, the final one being Header/Menu improvement (waterfall PR FINAL)  #596

W3stside commented on src/custom/hooks/useChangeNetworks.ts:
```diff
@@ -9,7 +9,7 @@ import usePrevious from 'hooks/usePrevious'
 import { useAppDispatch } from 'state/hooks'
 import { replaceURLParam } from 'utils/routes'
-import { getChainNameFromId, getParsedChainId } from 'components/Header/NetworkSelector'
+import { getChainNameFromId, getParsedChainId } from 'components/Header/NetworkSelector/NetworkSelectorMod'
```
any reason you changed this? if it's not exported in index could you re-export what u need there?

fairlighteth replied:
Yes will address

W3stside commented on src/custom/components/Header/MobileMenuIcon/index.tsx:
```diff
@@ -0,0 +1,82 @@
+import styled, { css } from 'styled-components/macro'
+
+const Wrapper = styled.div<{ isMobileMenuOpen: boolean; height?: number; width?: number; lineSize?: number }>`
+  z-index: 102;
```
102 😬😬😬

fairlighteth replied:
Yeah I know...I'd want to tackle the z-index thing in a separate PR. I think moving it to a const file, and then defining some general levels would be better.

W3stside commented on src/custom/components/Header/MobileMenuIcon/index.tsx:
```diff
@@ -0,0 +1,82 @@
+  margin: 0 6px 0 16px;
+  position: relative;
+  width: ${({ width }) => (width ? width + 'px' : '34px')};
+  height: ${({ height }) => (height ? height + 'px' : '18px')};
```
In these kinds of props i prefer setting the default in the function param destructuring like:
```ts
width: ${({ width = 18 }) => `${width}px`};
```

It just looks cleaner and is less code

fairlighteth replied:
Ah nice, of course this is way cleaner setting the default like that!

W3stside commented on src/custom/components/Header/MobileMenuIcon/index.tsx:
```diff
@@ -0,0 +1,82 @@
+  span {
+    background-color: ${({ theme }) => theme.text1};
+    border-radius: 3px;
+    height: ${({ lineSize }) => (lineSize ? lineSize + 'px' : '2px')};
```
I won't mention all of these instances but could also be replaced here

W3stside commented on src/custom/components/Header/MobileMenuIcon/index.tsx:
```diff
@@ -0,0 +1,82 @@
+
+export default function MobileMenuIcon({ isMobileMenuOpen, width, height, lineSize, onClick }: IconProps) {
+  return (
+    <Wrapper isMobileMenuOpen={isMobileMenuOpen} width={width} height={height} lineSize={lineSize} onClick={onClick}>
```
Since all the prop names are the same you can just not destructure the props in the function signature and pass them to the component like <Component {...params} />

W3stside commented on src/custom/components/MenuDropdown/index.tsx:
```diff
@@ -0,0 +1,38 @@
+export function Menu({ title, children }: MenuProps) {
+  const isUpToLarge = useMediaQuery(upToLarge)
+  const [showMenu, setShowMenu] = useState(false)
+  const handleOnClick = () => isUpToLarge && setShowMenu(!showMenu)
```
SetStates allow passing a callback as a param so it's safer to do:
```ts
setShowMenu(menuState => !menuState)
```

fairlighteth replied:
Like this ?
`const handleOnClick = () => isUpToLarge && setShowMenu((showMenu) => !showMenu)`

Why is it safer?

W3stside commented on src/custom/constants/routes.ts:
```diff
@@ -0,0 +1,24 @@
+// ENUM with routes
```
Woo nice good call!

W3stside commented on src/custom/utils/toggleBodyClass.ts:
```diff
@@ -0,0 +1,7 @@
+export const toggleBodyClass = (className: string, isAdd: boolean) => {
```
toggle in the name implies inverting the current state so the `isAdd` param is redundant

this function can just be written as 
```ts
export const toggleBodyClass = (className: string) => {
  if (!document.body.classList.contains(className)) {
    document.body.classList.add(className)
  } else {
    document.body.classList.remove(className)
  }
}

fairlighteth replied:
Makes sense yes. Actually I created also specific addBodyClass and removeBodyClass util functions, as that seems sometimes more convenient.

W3stside reviewed (approved):
approving with minor comments

elena-zh commented:
Hey @fairlighteth , great job!

Some issues I faced during testing: 
1. I noticed that some links remain to be underlined when navigate to another FAQ section: 
![image](https://user-images.githubusercontent.com/70885163/172362717-94c947a6-302a-4084-9d04-465a7775e463.png)
2. Seems like 'Protocol' section is missing in the FAQ now: 'Token' section is opened when press on the 'Protocol'
![image](https://user-images.githubusercontent.com/70885163/172362776-b506f0f2-0bff-4afc-9b9e-ff30a7d28e02.png)



3 FAQ overview link is always highlighted, even if another section is selected 
![image](https://user-images.githubusercontent.com/70885163/172363012-92acedec-5706-45a2-9f50-e152a2bfc096.png)

4. WDYT about auto-closing dropdowns after selecting this or that option? Currently, dropdown remains opened unless I remove a mouse from it

5. I can't open a menu in a tablet/mobile view (Win10+Chrome, FF, Brave)
![image](https://user-images.githubusercontent.com/70885163/172364496-ed22325f-abd2-4ddb-a224-e34be72d265a.png)

6. Maybe we should display feedback window above the feedback icon, not from the right side? WDYT?
![image](https://user-images.githubusercontent.com/70885163/172364811-9f6187be-34c1-4cff-b2cc-e85570cf184f.png)
At least, it will not overlap Connect wallet button

7. Footer in a tablet view hides contracts/block/gas price info
![image](https://user-images.githubusercontent.com/70885163/172365120-abc789c2-4a5a-43d5-a5a2-9954e237340c.png)
 
8. Connect to wallet button has a light boarder in the light mode
![image](https://user-images.githubusercontent.com/70885163/172366232-1def69a1-fc2e-49ee-b4db-9ec0fba91db3.png)

9. The issue with FAQ mobile sections' alignment #486 is also reproducible in this PR 
10. It is almost impossible to scroll down the menu in iPhone with iOS 15.3

https://user-images.githubusercontent.com/70885163/172367281-023e52e3-e39c-4dcc-840a-76bd8bd2949c.mp4

11. Account Affiliate section elements exceed the section's height  in iPhone 12 mini (when a user is not connected)
![image](https://user-images.githubusercontent.com/70885163/172367613-df3756ad-37df-4e8b-8970-f7e69e4ebbea.png)
in Android 'Connect wallet' button is shifted to the right

12. Maybe just like a reminder: the issue with the feedback window for iOS wallets integration browsers is still broken (#165)
13. And as an enhancement I'd like to propose to show COW and ETH balances at the top of the menu (however, do not have a strong opinion on this)
![image](https://user-images.githubusercontent.com/70885163/172369530-b3bec183-27a5-492a-987f-afb88931f3d4.png)


Thanks!

fairlighteth commented:
1. Yes, this is to indicate the current active page in the sub-menu. In this case I will change it to a different color instead
<img width="158" alt="Screen Shot 2022-06-07 at 12 44 14" src="https://user-images.githubusercontent.com/31534717/172371320-64f8f773-7b33-4292-8a74-3aa1b340f1f8.png">

2 and 3: Addressed this and point 1 now to better handle this.

4. Yes makes sense. I now made sure that on click of any menu element is closes the dropdown menu. On click anywhere outside the menu > it closes as well. Also instead of mouse-over, I made it to open the dropdown on click of the parent element (with the arrow).

Addressed these in https://github.com/cowprotocol/cowswap/pull/650

elena-zh commented:
> 5. I think what happens is that the onClick events aren't refreshed when you change viewport. Can you try again by refreshing the page? If it does work then we need to address that issue.

@fairlighteth , refreshing does not help :(
https://watch.screencastify.com/v/uuW5RvCrudYQR9kWMA81

fairlighteth commented:
@elena-zh Thanks for checking again. The borders were intentionally removed to allow for more horizontal space. By doing that, the container/content area has more available width to display content/numbers.

fairlighteth commented:
@elena-zh Actually I would leave that one in place.

elena-zh commented:
@fairlighteth , I have created: 
#668 : new issue
#669 : issue related to Account page Connect button 

Btw, I'm still not sure if it is OK that we have a white boarder for the Connect button in the light mode? 
![image](https://user-images.githubusercontent.com/70885163/172854102-89ffe699-b63d-431e-b0a8-f146bc1fed24.png)

Btw, where is Terms and Conditions page now?

Thanks

fairlighteth commented:
@elena-zh 

- Indeed forgot to add the terms and conditions -> https://github.com/cowprotocol/cowswap/pull/670
- When merging the Accounts page PR at some point, I will also check the 'view orders' menu item.
- In regard to the white border on the button, personally I think this is fine (intentional that is).

elena-zh commented:
Hey @fairlighteth , 
after the final round of testing I have reported some new/missing issues: 
 - https://github.com/cowprotocol/cowswap/issues/672 (new)
 - https://github.com/cowprotocol/cowswap/issues/673 (I have missed it)
 - https://github.com/cowprotocol/cowswap/issues/276#issuecomment-1152180943 - added a note to the existing issue. 

Besides, I still vote for fixing https://github.com/cowprotocol/cowswap/issues/669 in the near future. 

But for now, the current PR is approved. 

Thanks!
